core/workspace_svg.js

```diff
@@ -1478,6 +1498,60 @@ WorkspaceSvg.prototype.render = function() {
   this.markerManager_.updateMarkers();
 };
 
+/**
+ * Handles rendering of blocks added to queue for drag events and removes
+ *     itself from the change listeners and empties the queue on the completion
+ *     of a drag event.
+ * @param {Blockly.Events.Abstract} e the event to act on accordingly.
+ * @private
+ */
+WorkspaceSvg.prototype.renderBlocksOnDrag_ = function(e) {
+  if ([Blockly.Events.BLOCK_DRAG, Blockly.Events.BLOCK_DELETE]
+      .includes(e.type)) {
+    this.blockDragRenderQueue_.forEach((blockBumpPair) => {
+      const block = blockBumpPair.block;
+      if (block.workspace === this) { // block hasn't been disposed
+        block.render();
+        if (blockBumpPair.bumpNeighbours) {
+          block.bumpNeighbours();
+        }
+      }
+    });
+
+    if (e.type === Blockly.Events.BLOCK_DRAG && !e.isStart) {
+      this.blockDragRenderQueue_ = [];
+      this.removeChangeListener(this.renderOnDragListener_);
+      this.renderOnDragListener_ = null;
+    }
+  }
+};
+
+/**
+ * Add block to queue of blocks to be rendered on block drag events up to and
+ *     including the next block drag event concludes.
+ * @param {Blockly.Block} block the block to be added.
+ * @param {boolean=} addDescendants If true, adds all descendants of this block
+ *     to this queue.
+ * @param {boolean=} bumpNeighbours If true, calls bumpNeighbors on drag for
+ *     this block (and descendants if allDescendants is true).
+ */
+WorkspaceSvg.prototype.addToDragRenderQueue = function(block, addDescendants, bumpNeighbours) {
+  if (!this.isDragging()) {
+    return;
+  }
+
+  const blocks = (addDescendants ? block.getDescendants() : [block])
```

Comment:
I think you can be a lot more discerning about which blocks get rendered, which may have an impact on performance for this change. Specifically, you shouldn't need to add descendants.

- When a set of blocks explodes during a drag, it's not that the leaf blocks were rendered wrong. It's that the parent block's rerender failed to update connection locations or, more frequently, recalculated connection locations wrong--e.g. by doubling offsets or misplacing the origin.

- Rendering "bubbles up", meaning that rendering a block will re-render its parents but not its children. So if you render a block and then one of its children, the child will render once but the parent and all blocks above it on the stack will render twice.